### PR TITLE
[Merged by Bors] - feat: injectivity of continuous linear maps is an open condition

### PR DESCRIPTION
**Pull request by grunweg**
in finite-dimensional spaces over a complete field. This is used to showing that immersions are open.
NB: this result is false for general continuous maps (only surjectivity is open there).

From the sphere-eversion project.

----------

Placing the lemmas about normed groups is tricky (their natural place is Normed/Group/Basic, but they also require the topology on NNReal...). Better solutions welcome.

Aside: I don't see a good label for this; I would label with "baby differential topology" if I could.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/NormedSpace/FiniteDimension.lean [hunk 1]
```diff
@@ -231,6 +231,32 @@ theorem LinearMap.exists_antilipschitzWith [FiniteDimensional 𝕜 E] (f : E →
     exact ⟨_, e.nnnorm_symm_pos, e.antilipschitz⟩
 #align linear_map.exists_antilipschitz_with LinearMap.exists_antilipschitzWith
 
+open Function in
+/-- A `LinearMap` on a finite-dimensional space over a complete field
+  is injective iff it is anti-Lipschitz. -/
+theorem LinearMap.injective_iff_antilipschitz
+    [FiniteDimensional 𝕜 E] [CompleteSpace 𝕜] (f : E →ₗ[𝕜] F) :
```
Comment (j-loreaux) [inline Mathlib/Analysis/NormedSpace/FiniteDimension.lean:238]: There is already a `[CompleteSpace 𝕜]` hypothesis in context.
```suggestion
theorem LinearMap.injective_iff_antilipschitz [FiniteDimensional 𝕜 E] (f : E →ₗ[𝕜] F) :
```

File: Mathlib/Analysis/NormedSpace/FiniteDimension.lean [hunk 2]
```diff
@@ -231,6 +231,32 @@ theorem LinearMap.exists_antilipschitzWith [FiniteDimensional 𝕜 E] (f : E →
...
+  constructor
+  · rw [← LinearMap.ker_eq_bot]
+    exact f.exists_antilipschitzWith
+  · rintro ⟨K, -, H⟩
+    exact H.injective
+
+open Function in
+/-- Injectivity of continuous linear maps `E → F` is open,
+  if `E` is finite-dimensional over a complete field. -/
+theorem ContinuousLinearMap.isOpen_injective [FiniteDimensional 𝕜 E] [CompleteSpace 𝕜] :
```
Comment (j-loreaux) [inline Mathlib/Analysis/NormedSpace/FiniteDimension.lean:249]:
```suggestion
theorem ContinuousLinearMap.isOpen_injective [FiniteDimensional 𝕜 E] :
```

File: Mathlib/Analysis/NormedSpace/FiniteDimension.lean [hunk 3]
```diff
@@ -231,6 +231,32 @@ theorem LinearMap.exists_antilipschitzWith [FiniteDimensional 𝕜 E] (f : E →
...
+    [FiniteDimensional 𝕜 E] [CompleteSpace 𝕜] (f : E →ₗ[𝕜] F) :
+    Injective f ↔ ∃ K > 0, AntilipschitzWith K f := by
+  constructor
+  · rw [← LinearMap.ker_eq_bot]
+    exact f.exists_antilipschitzWith
+  · rintro ⟨K, -, H⟩
+    exact H.injective
+
+open Function in
+/-- Injectivity of continuous linear maps `E → F` is open,
```
Comment (j-loreaux) [inline Mathlib/Analysis/NormedSpace/FiniteDimension.lean:247]: This phrasing to me sounds a bit better.
```suggestion
/-- The set of injective continuous linear maps `E → F` is open,
```

File: Mathlib/Analysis/Normed/Group/Basic.lean [hunk 4]
```diff
@@ -9,6 +9,7 @@ import Mathlib.Topology.Instances.Rat
 import Mathlib.Topology.MetricSpace.Algebra
 import Mathlib.Topology.MetricSpace.IsometricSMul
 import Mathlib.Topology.Sequences
+import Mathlib.Topology.Instances.NNReal
```
Comment (j-loreaux) [inline Mathlib/Analysis/Normed/Group/Basic.lean:12]: This pulls in a *lot* of new dependencies, so we really shouldn't do this. `Analysis.Normed.Group.Basic` is too long already. Please start a new file `Analysis.Normed.Group.Lemmas` where you can put your new lemmas, and add a TODO to it saying we should split part of  `Analysis.Normed.Group.Basic` off into this file.

File: Mathlib/Analysis/NormedSpace/FiniteDimension.lean [hunk 5]
```diff
@@ -231,6 +231,32 @@ theorem LinearMap.exists_antilipschitzWith [FiniteDimensional 𝕜 E] (f : E →
...
+theorem ContinuousLinearMap.isOpen_injective [FiniteDimensional 𝕜 E] [CompleteSpace 𝕜] :
+    IsOpen { L : E →L[𝕜] F | Injective L } := by
+  rw [isOpen_iff_eventually]
+  rintro φ₀ hφ₀
+  rcases φ₀.injective_iff_antilipschitz.mp hφ₀ with ⟨K, K_pos, H⟩
+  have : ∀ᶠ φ in 𝓝 φ₀, ‖φ - φ₀‖₊ < K⁻¹ := eventually_nnnorm_sub_lt _ <| inv_pos_of_pos K_pos
+  filter_upwards [this] with φ hφ
+  apply φ.injective_iff_antilipschitz.mpr
+  exact ⟨(K⁻¹ - ‖φ - φ₀‖₊)⁻¹, inv_pos_of_pos (tsub_pos_of_lt hφ),
+    AntilipschitzWith.add_sub_lipschitzWith H (φ - φ₀).lipschitz hφ⟩
```
Comment (j-loreaux) [inline Mathlib/Analysis/NormedSpace/FiniteDimension.lean:258]: optional
```suggestion
    H.add_sub_lipschitzWith (φ - φ₀).lipschitz hφ⟩
```

**Review by j-loreaux — COMMENTED** [review]
I'll note that you can get your result directly without going directly through `AntiLipschitz` (although ultimately it is using it under the hood somewhere), in the following way:
```lean
theorem isOpen_setOf_clm_injective [FiniteDimensional 𝕜 E] :
    IsOpen { f : E →L[𝕜] F | Function.Injective f } := by
  suffices this : ∀ (f : E →L[𝕜] F), Function.Injective f ↔ finrank 𝕜 E ≤ (f : E →ₗ[𝕜] F).rank by
    simpa [this] using isOpen_setOf_nat_le_rank (finrank 𝕜 E)
  simp_rw [LinearMap.rank, ← finrank_eq_rank]
  refine fun f ↦ ⟨by exact_mod_cast (LinearMap.finrank_range_of_inj · |>.symm |>.le), ?_⟩
  norm_cast
  rw [← add_zero (finrank 𝕜 (LinearMap.range (f : E →ₗ[𝕜] F))),
    ← f.finrank_range_add_finrank_ker]
  exact fun hf ↦ (LinearMapClass.ker_eq_bot (E →L[𝕜] F)).mp <| by
    simpa using Nat.eq_zero_of_le_zero (le_of_add_le_add_left hf)
```
That being said, I think we want the other results in this PR anyway, so no need to change your proof.

File: Mathlib/Analysis/NormedSpace/FiniteDimension.lean (hunk 3 above)
Comment (grunweg) [reply Mathlib/Analysis/NormedSpace/FiniteDimension.lean]: I agree! "An open condition" is common, but I don't see a need to this jargon.

**Comment by grunweg** [comment]
Thank you for the useful comments. I have addressed them, so this is ready for another look.
awaiting-review

**Comment by j-loreaux** [comment]
@grunweg it's missing the new import

File: Mathlib/Analysis/Normed/Group/Lemmas.lean [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+-/
+import Mathlib.Analysis.Normed.Group.Basic
+import Mathlib.Topology.Instances.NNReal
+
+/-!
+## Further lemmas about normed groups
+
+This file contains further lemmas about normed groups, requiring heavier imports than
+`Mathlib/Analysis/Normed/Group/Basic.lean`.
+TODO(everybody): move lemmas from `Basic` to other places, included this.
```
Comment (j-loreaux) [inline Mathlib/Analysis/Normed/Group/Lemmas.lean:14]:
```suggestion
## TODO

- Move lemmas from `Basic` to other places, including this file.
```

**Comment by j-loreaux** [comment]
bors d+

**Comment by grunweg** [comment]
Indeed, thanks. Fixed now.
bors r+

**Pull request closed without merging**